### PR TITLE
feat: add notion of addressbooks

### PR DESCRIPTION
**Pull request by asbiin**
This adds the notion of Address Book
It's the first step for  #2738

This let an account have multiple address book.
The default one is the current solution (with only 1 address book).
The notion of address books is already integrated on carddav.

This PR only address the ability to *have* multiple address books.
Other PR will gave the opportunity to
- add other distant address books, and add subscription to them
- connect to carddav clients and update them regularly
- display address books content on dashboard

All API functions only address the default address book.

So basically it does not change a lot of things immediatly.

### CODE REVIEW

File: database/migrations/2020_03_25_055551_add_address_book.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            $table->foreign('user_id')->references('id')->on('users')->onDelete('cascade');
+        });
+    }
+
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
```
Comment (djaiss) [inline database/migrations/2020_03_25_055551_add_address_book.php]: We don't need down() method at all. We never rollback. I don't include them in my code ever.

**Comment by djaiss** [comment]
I don't understand this feature. Can you explain more? Also, I have the feeling it makes the code much more complex, for instance when downgrading, now, as a developer, we will need to think about checking contacts in the address book. Why? How can we make it very simple for developers to not think about that?

File: database/migrations/2020_03_25_055551_add_address_book.php (hunk 1 above)
Comment (asbiin) [reply database/migrations/2020_03_25_055551_add_address_book.php]: Well, I do.

**Comment by asbiin** [comment]
We could add parameters used on the scope function to the `Account::contact` one, so it will be easy for developers, because `account->contact()` will be the default address book, and `account->contact($id, $abid)` will return a specific contact list.
The only limit is to get the ALL contacts of an account. Maybe if addressbookId is `-1`, it return every contacts.

**Comment by tomperrine** [comment]
> I don't understand this feature. Can you explain more? Also, I have the feeling it makes the code much more complex, for instance when downgrading, now, as a developer, we will need to think about checking contacts in the address book. Why? How can we make it very simple for developers to not think about that?

For example, I have multiple address books - one from O365 at work, a personal gmail account, a consulting email account - I'd like to have all these in Monica, yet be able to manipulate them independently when needed.

File: app/Models/Contact/Contact.php [hunk 2]
```diff
@@ -162,6 +163,26 @@ public function account()
...
+    }
+
+    /**
+     * Get the list of contacts from the same address book as this contact.
+     *
+     * @return HasMany|null
+     */
+    public function addressBookContacts()
+    {
+        return $this->account ? $this->account->addressBookContacts($this->addressBook ? $this->addressBook->name : null) : null;
```
Comment (djaiss) [inline app/Models/Contact/Contact.php:183]: Way too complex in my opinion.
You are trying to be too smart 😀 Yes the code is more pure that way, but it comes with a price of having to decrypt the line to understand what it does.

Could you make it multi line so it’s easy to read in one glance?

File: app/Models/Contact/Contact.php [hunk 3]
```diff
@@ -162,6 +163,26 @@ public function account()
...
+    {
+        return $this->belongsTo(AddressBook::class);
+    }
+
+    /**
+     * Get the list of contacts from the same address book as this contact.
+     *
+     * @return HasMany|null
+     */
+    public function addressBookContacts()
```
Comment (djaiss) [inline app/Models/Contact/Contact.php:181]: I don't think this is a good name.

To reduce complexity we should have simple names. I think, in this case, what are you trying to achieve? To get the contacts from the same address book.
So it could be `similarContactsInAddressBook()` or `sameAddressBookContacts()`.

File: app/Models/Contact/Contact.php [hunk 4]
```diff
@@ -546,6 +567,28 @@ public function scopeNotActive($query)
         return $query->where('is_active', 0);
     }
 
+    /**
+     * Scope a query to only include contacts from designated address book.
```
Comment (djaiss) [inline app/Models/Contact/Contact.php:571]:
```suggestion
     * Scope a query to only include contacts from given address book.
```

File: app/Http/Controllers/DashboardController.php [hunk 5]
```diff
@@ -28,19 +28,17 @@ public function index()
             )->with('debts.contact')
             ->first();
 
-        $numberOfContacts = $account->contacts()
-                                    ->real()
-                                    ->active()
-                                    ->count();
+        $numberOfContacts = $account->addressBookContacts()
```
Comment (djaiss) [inline app/Http/Controllers/DashboardController.php:31]: Here I don't like it 😀
Why do we have to think about querying this method instead of `contacts`?

I think address books should be transparent unless we really want to manage it in the code/controller because we need to manage different address books.

In this case, the right call should be `account->contacts()` and that's it. We should manage the notion of address books in the `contacts()` method, and that way everything would be transparent.

Or perhaps it should be something like 

```
$account->contacts()
     ->real()
     ->active()
     ->count();
```

and we could have another scope to specify the address book when we need to scope this to a specific address book?

File: database/migrations/2020_03_25_201407_add_contact_vcard_data.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('contacts', function (Blueprint $table) {
+            $table->string('vcard', 65536)->after('gravatar_url')->nullable();
```
Comment (djaiss) [inline database/migrations/2020_03_25_201407_add_contact_vcard_data.php:17]: What does this do? It’s not clear.

File: app/Services/VCard/ImportVCard.php [hunk 7]
```diff
@@ -414,7 +439,12 @@ private function existingContactWithEmail(VCard $entry): ?Contact
                 'contact_field_type_id' => $this->getContactFieldTypeId(ContactFieldType::EMAIL),
             ])->whereIn('data', iterator_to_array($entry->EMAIL))->first();
 
-            if ($contactField) {
+            if ($contactField &&
```
Comment (djaiss) [inline app/Services/VCard/ImportVCard.php:442]: We should add a comment above this `if` to give the reader an overview of what this complex `if` does.

File: app/Http/Controllers/DashboardController.php (hunk 5 above)
Comment (asbiin) [reply app/Http/Controllers/DashboardController.php:31]: That was my first try.. and it does not work like this. contacts() return a HasMany, and I can add a filter in this function. I don't know where to insert the scope/Query without explicitly call it. If you have a better idea ok, but I have zero.

File: database/migrations/2020_03_25_201407_add_contact_vcard_data.php (hunk 6 above)
Comment (asbiin) [reply database/migrations/2020_03_25_201407_add_contact_vcard_data.php:17]: The best explanation is here:
https://sabre.io/dav/building-a-carddav-client/#retain-full-vcards%21

Comment (asbiin) [reply database/migrations/2020_03_25_201407_add_contact_vcard_data.php]: The idea is to record the generated vcard.
It will only be used for external address books. Internal default address book will not save the vcard.

File: app/Http/Controllers/DashboardController.php (hunk 5 above)
Comment (asbiin) [reply app/Http/Controllers/DashboardController.php:31]: There is not "less complex code" where we code magically guess the address book currently seeing.

I guess I could just revert the changes and go back to something like
```
$account->contacts()
     ->addressbook($addressBookName)
     ->real()
     ->active()
     ->count();
```
If $addressBookName is null it will scope to default address book (null id), if not it will scope to the address book.

**Comment by asbiin** [comment]
Thanks for the review.
I don't think it adds so complex code .
However, I understand your point, and I tried to simplify it. I guess it's not working for you ...
